core/executor/src/wasmtime/runtime.rs

```diff
@@ -0,0 +1,370 @@
+	let host_state = TrampolineState::new::<SubstrateExternals>(code_memory);
+
+	let result = InstanceHandle::new(
+		Rc::new(module),
+		global_exports,
+		finished_functions.into_boxed_slice(),
+		imports,
+		&data_initializers,
+		signatures.into_boxed_slice(),
+		None,
+		Box::new(host_state),
+	);
+	result.map_err(|e| WasmError::WasmtimeSetup(SetupError::Instantiate(e)))
+}
+
+/// Build a new TargetIsa for the host machine.
+fn target_isa() -> std::result::Result<Box<dyn TargetIsa>, WasmError> {
+	let isa_builder = cranelift_native::builder()
+		.map_err(WasmError::MissingCompilerSupport)?;
+	let flag_builder = cranelift_codegen::settings::builder();
+	Ok(isa_builder.finish(cranelift_codegen::settings::Flags::new(flag_builder)))
+}
+
+fn clear_globals(global_exports: &mut HashMap<String, Option<Export>>) {
+	global_exports.remove("memory");
+	global_exports.remove("__heap_base");
+	global_exports.remove("__indirect_function_table");
+}
+
+unsafe fn grow_memory(instance: &mut InstanceHandle, pages: u32) -> Result<()> {
+	let memory_index = match instance.lookup_immutable("memory") {
+		Some(Export::Memory { definition, vmctx: _, memory: _ }) =>
+			instance.memory_index(&*definition),
+		_ => return Err(Error::InvalidMemoryReference),
+	};
+	instance.memory_grow(memory_index, pages)
+		.map(|_| ())
+		.ok_or_else(|| "requested heap_pages would exceed maximum memory size".into())
+}
+
+fn get_host_state(context: &mut Context) -> Result<&mut TrampolineState> {
+	let env_instance = context.get_instance("env")
+		.map_err(|err| format!("cannot find \"env\" module: {}", err))?;
+	env_instance
+		.host_state()
+		.downcast_mut::<TrampolineState>()
+		.ok_or_else(|| "cannot get \"env\" module host state".into())
+}
+
+fn reset_host_state(context: &mut Context, executor_state: Option<FunctionExecutorState>)
+	-> Result<Option<Error>>
+{
+	let trampoline_state = get_host_state(context)?;
+	trampoline_state.executor_state = executor_state;
+	Ok(trampoline_state.reset_trap())
+}
+
+unsafe fn inject_input_data(
+	context: &mut Context,
+	instance: &mut InstanceHandle,
+	data: &[u8],
+) -> Result<(Pointer<u8>, WordSize)> {
+	let trampoline_state = get_host_state(context)?;
+	let executor_state = trampoline_state.executor_state
+		.as_mut()
+		.ok_or_else(|| "cannot get \"env\" module executor state")?;
+
+	let memory = get_memory_mut(instance)?;
+
+	let data_len = data.len() as WordSize;
+	let data_ptr = executor_state.heap().allocate(memory, data_len)?;
+	write_memory_from(memory, data_ptr, data)?;
+	Ok((data_ptr, data_len))
+}
+
+unsafe fn get_memory_mut(instance: &mut InstanceHandle) -> Result<&mut [u8]> {
+	match instance.lookup("memory") {
+		Some(Export::Memory { definition, vmctx: _, memory: _ }) =>
+			Ok(std::slice::from_raw_parts_mut(
+				(*definition).base,
+				(*definition).current_length,
+			)),
+		_ => Err(Error::InvalidMemoryReference),
+	}
+}
+
+unsafe fn get_heap_base(instance: &InstanceHandle) -> Result<u32> {
```

Comment:
It also dereferences the `definition` pointer. That is looked up from an instance, however, so maybe it is all right to confine the `unsafe` ness. On the other hand, I don't really see the point as it's only called from within an unsafe block.

Reply:
> On the other hand, I don't really see the point as it's only called from within an unsafe block.

So that's the point : ) My objective is to minimize and confine `unsafe`ness to be little islands in the ocean of safeness rather than the other way around. It would be better if we removed that unsafe block altogether.

The best would be if we could manage to keep unsafe invariants within a single module, although it might be a pipe dream.